src/token/cose/encrypt/encrypt0.rs

```diff
@@ -0,0 +1,553 @@
+        CKP: CoseKeyProvider<'a>,
+        CAP: CoseAadProvider<'b>,
+    >(
+        &self,
+        backend: &mut B,
+        key_provider: &mut CKP,
+        try_all_keys: bool,
+        external_aad: &mut CAP,
+    ) -> Result<Vec<u8>, CoseCipherError<B::Error>>;
+}
+
+impl CoseEncrypt0Ext for CoseEncrypt0 {
+    fn try_decrypt<
+        'a,
+        'b,
+        B: CoseEncryptCipher,
+        CKP: CoseKeyProvider<'a>,
+        CAP: CoseAadProvider<'b>,
+    >(
+        &self,
+        backend: &mut B,
+        key_provider: &mut CKP,
+        try_all_keys: bool,
+        external_aad: &mut CAP,
+    ) -> Result<Vec<u8>, CoseCipherError<B::Error>> {
+        self.decrypt(
+            external_aad.lookup_aad(Some(&self.protected.header), Some(&self.unprotected)),
+            |ciphertext, aad| {
+                try_decrypt(
+                    backend,
+                    key_provider,
+                    &self.protected.header,
+                    &self.unprotected,
+                    try_all_keys,
+                    ciphertext,
+                    aad,
+                )
+            },
+        )
+    }
+}
+
+pub trait CoseEncrypt0BuilderExt: Sized {
+    fn try_encrypt<
+        'a,
+        'b,
+        B: CoseEncryptCipher,
+        CKP: CoseKeyProvider<'a>,
+        CAP: CoseAadProvider<'b>,
+    >(
+        self,
+        backend: &mut B,
+        key_provider: &mut CKP,
+        try_all_keys: bool,
+        protected: Option<Header>,
+        unprotected: Option<Header>,
+        plaintext: &[u8],
+        external_aad: &mut CAP,
+    ) -> Result<Self, CoseCipherError<B::Error>>;
+}
+
+impl CoseEncrypt0BuilderExt for CoseEncrypt0Builder {
+    fn try_encrypt<
+        'a,
+        'b,
+        B: CoseEncryptCipher,
+        CKP: CoseKeyProvider<'a>,
+        CAP: CoseAadProvider<'b>,
+    >(
+        self,
+        backend: &mut B,
+        key_provider: &mut CKP,
+        try_all_keys: bool,
+        protected: Option<Header>,
+        unprotected: Option<Header>,
+        plaintext: &[u8],
+        external_aad: &mut CAP,
+    ) -> Result<Self, CoseCipherError<B::Error>> {
+        let mut builder = self;
+        if let Some(protected) = &protected {
+            builder = builder.protected(protected.clone());
+        }
+        if let Some(unprotected) = &unprotected {
+            builder = builder.unprotected(unprotected.clone());
+        }
+        builder.try_create_ciphertext(
+            plaintext,
+            external_aad.lookup_aad(protected.as_ref(), unprotected.as_ref()),
+            |ciphertext, aad| {
+                try_encrypt_single(
+                    backend,
+                    key_provider,
+                    protected.as_ref(),
+                    unprotected.as_ref(),
+                    try_all_keys,
+                    ciphertext,
+                    aad,
+                )
+            },
+        )
+    }
+}
+
+#[cfg(all(test, feature = "std"))]
+mod tests {
```

Comment:
To stay consistent with the rest of `dcaf-rs`, test modules should be moved into their own file.